### PR TITLE
Implement ExecutionTracingProvider

### PR DESCRIPTION
**Pull request by jvmncs**
Implements a TracingProvider that reproduces the logging from tf-encrypted/rfcs#23 etc.

Currently the Bazel build isn't working, so I can't test it on a real TFF computation. I imagine I will have some debugging work to do once that's fixed.

### CODE REVIEW

File: tensorflow_federated/python/core/utils/execution_tracing.py
Comment (mortendahl) [inline tensorflow_federated/python/core/utils/execution_tracing.py]: extra line break

File: tensorflow_federated/python/core/utils/execution_tracing.py
Comment (mortendahl) [inline tensorflow_federated/python/core/utils/execution_tracing.py]: perhaps reformat comment to match elsewhere

File: tensorflow_federated/python/core/utils/BUILD
Comment (mortendahl) [inline tensorflow_federated/python/core/utils/BUILD:120]: problem is that you're not using this target as a dependency anywhere (bazel will only include what's needed, e.g. dependencies)

**Review by mortendahl — COMMENTED** [review]
looks good! pushed a quick fix for the missing file.

quick snippet for how to test it? or perhaps an unittest?

File: tensorflow_federated/python/core/utils/execution_tracing.py
Comment (jvmncs) [reply tensorflow_federated/python/core/utils/execution_tracing.py]: this was matching the original RFC file, but I agree

File: tensorflow_federated/python/core/utils/BUILD
Comment (jvmncs) [reply tensorflow_federated/python/core/utils/BUILD:120]: :dizzy_face:

**Comment by jvmncs** [comment]
I added two tests, I still need a test for the actual tracing logic but all else seems to be working properly now. The last test is currently blocked by figuring out how to use python logging to read from the file it's currently writing to -- if anyone @tf-encrypted/maintainers knows how to do this please ping me! I'd also be okay with merging as is after a quick spot test with a simpler executor stack.

I found one problem -- when using the `ThreadDelegatingExecutor`, things get logged out of order (as expected). Correlating this seems beyond our current goals, so I'm inclined to do one of the following:
- allow for the weird behavior by just ignoring it,
- log a warning upon detecting a ThreadDelegatingExecutor, or
- raise an exception upon detecting a ThreadDelegatingExecutor.

**Comment by mortendahl** [comment]
> blocked by figuring out how to use python logging to read from the file it's currently writing to ...  I'd also be okay with merging as is after a quick spot test with a simpler executor stack.

Reading from files seems delicate. I'd be okay with merging as is as well.

> when using the ThreadDelegatingExecutor, things get logged out of order

I think `allow for the weird behavior by just ignoring it` is a good option for now if you add a note in the class docstring.

**Review by mortendahl — APPROVED** [review]
lgtm

**Comment by jvmncs** [comment]
Currently, we have these options:
- TraceFilterMode.VALUES: traced functions must match `create_([a-z]+` to be logged
- TraceFilterMode.INTRINSICS: traced functions must match `create_([a-z]+)` or `_compute_intrinsic_([a-z_]+)` to be logged
- TraceFilterMode.ALL: no pattern matching, log all traced functions

This should be easy to extend if we feel we need more flexibility.

I tested the following script with each possible value of `trace_filter_mode` to ensure that filtering logic works properly.

```python
import logging

import silence_tensorflow.auto
import tensorflow as tf
import tensorflow_federated as tff
from tensorflow_federated.python.common_libs import tracing
from tensorflow_federated.python.core.utils import execution_tracing

logging.basicConfig(format="%(message)s")
logger = logging.getLogger()
logger.setLevel(logging.DEBUG)

tracer = execution_tracing.ExecutionTracingProvider(
    trace_filter_mode=execution_tracing.TraceFilterMode.ALL)
tracing.set_tracing_provider(tracer)

@tff.tf_computation(tf.float32)
def add_half(x):
  return tf.add(x, 0.5)

@tff.federated_computation(tff.FederatedType(tf.float32, tff.CLIENTS))
def add_half_on_clients(x):
  return tff.federated_map(add_half, x)

def executor_stack_fn(mapping):
  del mapping

  def _make_test_executor(
      num_clients,
      use_reference_resolving_executor,
  ) -> tff.framework.FederatingExecutor:
    def bottom_ex():
      eager = tff.framework.EagerTFExecutor()
      if not use_reference_resolving_executor:
        return eager
      return tff.framework.ReferenceResolvingExecutor(eager)

    fed_ex = tff.framework.FederatingExecutor({
        tff.SERVER: bottom_ex(),
        tff.CLIENTS: [bottom_ex() for _ in range(num_clients)],
        None: bottom_ex(),
    })
    if use_reference_resolving_executor:
      return tff.framework.ReferenceResolvingExecutor(fed_ex)
    return fed_ex

  return _make_test_executor(
      num_clients=3, use_reference_resolving_executor=True)

factory = tff.framework.create_executor_factory(executor_stack_fn)
tff.framework.set_default_executor(factory)

res = add_half_on_clients([1.5, 2.5, 3.5])
print(res)
```

And the result is attached here: [log.txt](https://github.com/tf-encrypted/federated/files/4429387/log.txt). Limiting to either of TraceFilterMode.VALUES or TraceFilterMode.INTRINSICS cleans this up considerably.

**Pull request merged**
